qi-sdk/profile/nonlocal/qi/main.rkt

```diff
@@ -54,8 +56,191 @@
+;;       ['done 'done]
+;;       [(cons 'skip new-state) (cons 'skip new-state)]
+;;       [(list 'yield value new-state)
+;;        (list 'yield (sqr value) new-state)]))
+;;   (let ([s (stream next-map-stream lst)])
+;;     (stream->list s)))
+
+;; case of case
+;; when there is a conditional based on the return value of a conditional
+;; invert which conditional is checked first
+;; (define (filter-map lst)
+;;   (define (next-list->stream state)
+;;     (cond [(null? state) 'done]
+;;           [else (list 'yield (car state) (cdr state))]))
+;;   (define (next-filter-stream state)
+;;     (cond [(null? state) (match 'done
+;;                            ['done 'done]
+;;                            [(cons 'skip new-state) (cons 'skip new-state)]
+;;                            [(list 'yield value new-state)
+;;                             (if (odd? value)
+;;                                 (list 'yield value new-state)
+;;                                 (cons 'skip new-state))])]
+;;           [else (match (list 'yield (car state) (cdr state))
+;;                   ['done 'done]
+;;                   [(cons 'skip new-state) (cons 'skip new-state)]
+;;                   [(list 'yield value new-state)
+;;                    (if (odd? value)
+;;                        (list 'yield value new-state)
+;;                        (cons 'skip new-state))])]))
+;;   (define (next-map-stream state)
+;;     (match (next-filter-stream state)
+;;       ['done 'done]
+;;       [(cons 'skip new-state) (cons 'skip new-state)]
+;;       [(list 'yield value new-state)
+;;        (list 'yield (sqr value) new-state)]))
+;;   (let ([s (stream next-map-stream lst)])
+;;     (stream->list s)))
+
+;; partially evaluate match on known argument
+(define (filter-map lst)
+  (define (next-list->stream state)
+    (cond [(null? state) 'done]
+          [else (list 'yield (car state) (cdr state))]))
+  (define (next-filter-stream state)
+    (cond [(null? state) (match 'done
+                           ['done 'done]
+                           [(cons 'skip new-state) (cons 'skip new-state)]
+                           [(list 'yield value new-state)
+                            (if (odd? value)
+                                (list 'yield value new-state)
+                                (cons 'skip new-state))])]
+          [else (match (list 'yield (car state) (cdr state))
+                  ['done 'done]
+                  [(cons 'skip new-state) (cons 'skip new-state)]
+                  [(list 'yield value new-state)
+                   (if (odd? value)
+                       (list 'yield value new-state)
+                       (cons 'skip new-state))])]))
+  (define (next-map-stream state)
+    (match (next-filter-stream state)
+      ['done 'done]
+      [(cons 'skip new-state) (cons 'skip new-state)]
+      [(list 'yield value new-state)
+       (list 'yield (sqr value) new-state)]))
+  (let ([s (stream next-map-stream lst)])
+    (stream->list s)))
```

Comment:
Maybe I'm missing something, but I don't see a partially-evaluated match. I think I would have expected the following:
```racket
;; partially evaluate match on known argument
(define (filter-map lst)
  (define (next-list->stream state)
    (cond [(null? state) 'done]
          [else (list 'yield (car state) (cdr state))]))
  (define (next-filter-stream state)
    (cond [(null? state) 'done]
          [else
            (let ([value (car state)]
                  [new-state (cdr state)])
              (if (odd? value)
                (list 'yield value new-state)
                (cons 'skip new-state)))]))
  (define (next-map-stream state)
    (match (next-filter-stream state)
      ['done 'done]
      [(cons 'skip new-state) (cons 'skip new-state)]
      [(list 'yield value new-state)
       (list 'yield (sqr value) new-state)]))
  (let ([s (stream next-map-stream lst)])
    (stream->list s)))
```

It also might be more efficient to collapse the "identity" cases in `match` to a default `[x x]` clause (where the result is the same as whatever is matched: in the `cons` case, this should avoid extra allocation unless the Racket or match compiler already notices this).

Reply:
I think we might have left off midway on this one. Thank you, this will make a great starting point for next time! To put this code in context, our immediate next step is to continue optimizing and attempt to reach the baseline performance we got with `fold`. If we get there, that will tell us the entire set of optimizations that would be needed for stream fusion to be efficient, although it seems unlikely that we'd do all these optimizations at the Qi compiler level. It would be interesting data, anyway, and has been fun to go through so far :)

Even if we don't end up using this, there's a chance that it could be useful to do as a first step before rewriting to fold (as mentioned in the last paragraph of [Outlook](https://github.com/drym-org/qi/wiki/Qi-Compiler-Sync-Aug-4-2023#outlook) in last week's notes).

Reply:
While I'm here: how do I run the relevant benchmarks for the filter-map/stream fusion? Is it `make something`, and can it run only this benchmark or do I have to let them all run?

If I write my own, I'd like to be able to compare. It looks like right now you comment out one implementation and add the next one, is that right?

Reply:
Sorry, just saw this. Yes it's `make profile-competitive`, and we just replace the "Qi" `filter-map` implementation in `qi-sdk/profile/nonlocal/qi/main.rkt` with whatever optimized version we want to test.

Unfortunately I don't think we can pick specific benchmarks for the "competitive" ones at the moment.

Reply:
Oh actually scratch that! We can do this:
```
$ cd qi-sdk/profile/nonlocal
$ ./report-competitive.rkt -s "filter-map" -s "filter-map (large list)"
```
That's good to know, this will save us time next time 😆